### ISSUE TITLE
Feedback regarding "Clickable elements must be focusable and should have interactive semantics." metric

### ISSUE THREAD

**Issue by 6TELOIV** [mdn/content#45144]
### MDN URL

https://developer.mozilla.org/en-US/docs/Web/Accessibility/Guides/Understanding_WCAG/Keyboard

### What specific section or headline is this issue about?

Clickable elements must be focusable and should have interactive semantics.

### What information was incorrect, unhelpful, or incomplete?

This metric is described as always required, no exceptions. But, this isn't always the case. I often use click handlers to make a larger containing element clickable and have it activate a default action for that item which is already present on a button/link child. Examples include cards with a link and table rows which are selectable to open an editing UI.

In both of these cases, it would be worse for me to change the containing element to be keyboard focusable, as it would be duplicative of the existing (focusable) action that it's acting redundant for.

For example, this UI is flagged, because I'm using click events to make the entire "ticket" card clickable to go to the "More Info" link, but the "More Info" link is alread focusable.

<img width="999" height="302" alt="Image" src="https://github.com/user-attachments/assets/659628d0-efb9-4fab-b5b2-dcb5de557ac6" />

### What did you expect to see?

The explanation should explain the exceptions (redundancy), and possibly it should be a warning (may be, vs should) in the UI.

### Do you have any supporting links, references, or citations?

<dl>
<dt><a href="https://design-system.w3.org/components/cards.html#block-link-cards">W3C block link cards</a></dt>
<dd>Similar to the example I've shown; using click handler event to make a card clickable as a link without breaking highlighting and interacting with content on the card.</dd>
<dt><a href="https://www.w3.org/WAI/WCAG22/Understanding/keyboard#examples">WCAG 2.2 SC 2.1.1 Example 7</a></dt>
<dd>Specifically example 7, talks about a submit button <em>not</em> being focusable, and that being OK because the form can be submitted from within the input field.</dd>
</dl>

### Do you have anything more you want to share?

Sure, I would like to share that I hope you are having a lovely day, whoever is reading this!

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/web/accessibility/guides/understanding_wcag/keyboard`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/Accessibility/Guides/Understanding_WCAG/Keyboard
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/accessibility/guides/understanding_wcag/keyboard/index.md
* Last commit: https://github.com/mdn/content/commit/7ba6358a0ff684cc67c60b76d6d972722bbf0d18
* Document last modified: 2025-12-17T15:24:11.000Z

</details>

**Comment by Josh-Cena** [comment]
@mdn/accessibility FYI.

I don't understand this page at all; which part of WCAG does it map to? Almost none of the guidelines it contains seem to be in WCAG.

**Comment by 6TELOIV** [comment]
The title and context of the page is quite confusing; this page is linked from the Accessibility Tools in Firefox for additional help on accessibility issues flagged; because accessibility is complex and semantic, certain automatic checks (such as this one) will have false positives sometimes, and of course can't fully cover WCAG.

I'm trying to guess at author intent, but it seems like the goal is to ground the test in the specific part(s) of WCAG the test is checking for (in this case, 2.1 Keyboard Accessible). But overall, it's quite confusing, because there are four pages on Principles (Perceivable, Operable, Understandable, Robust), one on a guideline (Keyboard Accessible), and one that doesn't match anything specific in WCAG (Text labels and names).

I think a better way to structure this would be to just have pages for each test name, and link back to the specific WCAG SC or Guideline that it talks about.

IMO all of that is out of scope of this Issue which has to do specifically with the "Clickable elements must be focusable and should have interactive semantics." having false positives and not clarifying as such.

**Comment by estelle** [comment]
Anytime you create a clickable element, that element must also be focusable.  Focusable elements should have interactive semantics.

If i understand the example correctly, there is an interactive element nested inside another interactive element?  interactive controls should not be nested; elements like buttons or links should not contain other focusable elements. This issue is more related to WCAG 4.1.2 Level A for keyboards and AT. Fix interactive elements nested inside each other (a button in a link) by making them siblings, not parent and child. While this appears to be a keyboard / AT issue, it is also a cognitive one. How is the sighted user to know clicking the card will take them to "more info"?

Closing

**Comment by 6TELOIV** [comment]
This example is a clickable card pattern; the body of the card is the same control as the main action (link in this case). Sighted users can tell this because hovering the card highlights the "More Info" link, and hovering "buy tickets" underlines it's link. Keyboard sighted users see a focus ring around the two appropriate links, and Screen Reader users hear "More Info" and "buy tickets" as they focus the two elements (there is more sr-only text to make the links discernible).

If you were to make the card focusable, you would have nested controls and that would be bad for usability. Additionally, it is not recommended to make entire cards links because the link text is excessively long. Also, screen reader users would now hear 3 things, 2 of which are the exact same think.

The "standard" pattern used for this is adding a ::before element to the link and positioning it over the entire card (and moving other interactive elements on top of that to remain interactive). The problem with this is that you can no longer highlight text inside the card.

That is why this event based pattern is recommended, and why I am saying that the test in Mozilla is too aggressive. [2.1.1](https://www.w3.org/TR/WCAG21/#keyboard) specifically speaks to this in the note:

> This does not forbid and should not discourage providing mouse input or other input methods in addition to keyboard operation.

Additionally, it speaks to [functionality](https://www.w3.org/TR/WCAG21/#dfn-functionality), not individual HTML elements. For this example, all functionality is available to keyboard users.

[4.1.2](https://www.w3.org/TR/WCAG21/#name-role-value) also speaks to [UI Components](https://www.w3.org/TR/WCAG21/#dfn-user-interface-components), _not_ HTML elements. In this example, the elements make up two UI components (the two links) and so both are represented by name, role and value (because they are simply `<a>` elements). Adding a third UI component to screen readers would be wrong, as there is no third UI component here.

I want to reiterate, I get why the test in Firefox works this way and I get why the article exists to explain it. It is not programmatically determinable if multiple elements make up one "UI component" because that is a semantic decision.

I'm asking that this is discussed on the docs so that people don't make something less accessible by adding more redundant focus targets, slowing keyboard and SR UI access down.

**Comment by 6TELOIV** [comment]
@estelle I wrote that reply on my phone, and I just realized I missed one of your specific points/questions:

> If i understand the example correctly, there is an interactive element nested inside another interactive element?

No; in the HTML there are two anchors that are siblings. For usability on touch screen devices where it may be unclear what the touch target is, it is desirable that the entire card is clickable rather than just the title link or a link in the footer/actions area of the card. You can see closer what the markup looks like at the [CSS Tricks article](https://css-tricks.com/block-links-the-search-for-a-perfect-solution/#method-4-sprinkle-javascript-on-the-second-method) referenced in the [WCAG Block link cards](https://design-system.w3.org/components/cards.html#block-link-cards) component.

**Comment by Josh-Cena** [comment]
I'll reopen this if there's no agreement that the current content is correct by Friday. Because this is in W3C's official examples, there's surely enough correctness about it, and the current content does not seem to be backed by any WCAG guidelines.

**Comment by Josh-Cena** [comment]
Reopening; I think there's something in our docs legitimately incompatible with the W3C design system here.
